storage-macros/src/lib.rs

```diff
@@ -0,0 +1,127 @@
+                }
+
+                let mut file = file.take(self.chunk_size);
+
+                let bytes_sent = tokio::io::copy(&mut file, send).await?;
+
+                if bytes_sent < self.chunk_size {
+                    anyhow::bail!("Chunk read to few bytes");
+                }
+
+                Ok(())
+            }
+
+            async fn recv_payload(recv: &mut RecvStream) -> Result<Self> {
+                let len = recv.read_u32().await?;
+
+                let mut buffer = vec![0u8; len as usize];
+                recv.read_exact(&mut buffer).await?;
+                let mut payload: Self = bincode::deserialize(&buffer)?;
+
+                let data_len = payload.chunk_size;
+
+                payload.data = TMP_STORAGE_ROOT
+                    .get()
+                    .expect("Temporary storage not initialized via config")
+                    .join(payload.chunk_id.to_string());
+
+                let file = tokio::fs::File::create(&payload.data).await?;
+                file.set_len(data_len).await?;
+                let mut writer = tokio::io::BufWriter::with_capacity(payload.chunk_size as usize, file);
+                let mut limited_recv = recv.take(data_len);
+                tokio::io::copy(&mut limited_recv, &mut writer).await?;
+                writer.flush().await?;
+
+                writer.into_inner().sync_all().await?;
+
+                Ok(payload)
+            }
+        }
+
+        impl Drop for #name {
+            fn drop(&mut self) {
+                if self.data.exists() {
+                    let _ = std::fs::remove_file(&self.data);
+                }
+            }
+        }
+    };
+    TokenStream::from(expanded)
+}
+
+#[proc_macro_derive(Message)]
+pub fn derive_message_payload_enum(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+    let name = &input.ident;
+
+    let (send_arms, recv_arms) = match input.data {
+        Data::Enum(DataEnum { variants, .. }) => {
+            let mut send_arms = Vec::new();
+            let mut recv_arms = Vec::new();
+
+            for (idx, variant) in variants.iter().enumerate() {
+                let variant_idx = idx as u8;
+                let variant_name = &variant.ident;
+
+                if let Fields::Unnamed(fields) = &variant.fields {
+                    if fields.unnamed.len() == 1 {
+                        let payload_ty = &fields.unnamed[0].ty;
+                        send_arms.push(quote! {
+                            #name::#variant_name(payload) => {
+                                send.write_u8(#variant_idx).await?;
+                                payload.send_payload(send).await?
+                            }
+                        });
+
+                        recv_arms.push(quote! {
+                            #variant_idx => Ok(#name::#variant_name(#payload_ty::recv_payload(recv).await?)),
+                        });
```

Comment:
:warning: Unhygienic macro! Please make sure to always fully qualify types in quoted code!